[BEAM-378] integrate setuptools in Maven build

wikier opened the pull request:
This PR provide an initial integration of the Python SDK in the Maven build, relying on the `exec-maven-plugin` to call `setuptools` ([BEAM-378](https://issues.apache.org/jira/browse/BEAM-378)). 

Notice I keep out `install` on purpose, because I guess it may need discussion.

---

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

wikier commented:
Regarding CI, ASF Jenkins would need some extra configuration; but I guess Travis build wouldn't work until they implement travis-ci/travis-ci#4090.

davorbonaci commented:
Thanks Sergio!

@silviulica, @aaltay, what do you think?

silviulica commented:
Thanks Sergio! Any help on doing Maven stuff is more than welcome. We need to keep the tests passing though:
1) setup.py contains a reference to apache_beam.get_version() this will not work since the file itself is responsible to install the apache_beam package. This is the reason the code in there is loading manually the version.py file and executing it. You will need to do something similar.
2) Most tests should start with doing a 'pip install apache_beam...' or more likely 'python setup.py test' on the source code. So we should clarify that in the Maven install section. This is probably what you meant by leaving out the install.
3) keeping the version of a Python package in pom.xml files seems very non-Pythonic to me. If there is precedence for it then so be it but otherwise I would prefer to have it in the version.py file  and perhaps reference it (if possible) in the pom.xml. Or keep it in two places (not-ideal but may fly for a while).
4) Also need to decide on a versioning naming/numbering scheme.
5) setup.py will have to contain the pom.xml as a package data (and get it installed) because the get_version() function is called at runtime.

I suggest as a way to make progress to evolve this CL into  setting up the Maven infrastructure w/o doing changes to the version.py machinery: add the pom.xml file, add setup.py code to install it with the package so get_version() works at runtime, corrections on names.

aaltay commented:
Hi Sergio,

> Regarding CI, ASF Jenkins would need some extra configuration; but I guess Travis build wouldn't work until they implement travis-ci/travis-ci#4090.

Travis should work currently and it is already running Python tests. We do not need the above bug to be fixed. Because python is already installed in every Travis worker image already (https://docs.travis-ci.com/user/ci-environment/#Runtimes) And we also run an apt-get install to install 2.7 version explicitly.

aaltay commented on .travis.yml:
```diff
@@ -32,6 +32,8 @@ addons:
     packages:
     - python2.7
+    - python-setuptools
+    - python-dill
```
This change should not require of adding additional packages here. setuptools is already part of the Python distribution for  Python 2 >=2.7.9 and setup.py requires (will install) dill.

aaltay commented on sdks/python/apache_beam/version.py:
```diff
@@ -17,4 +17,13 @@
+    elements = pom.xpath(r'/pom:project/pom:parent/pom:version', namespaces={'pom':'http://maven.apache.org/POM/4.0.0'})
+    version = elements[0].text
+    version = version.replace("-SNAPSHOT", ".dev")
+    # TODO: PEP 440 and incubatign suffix
```
incubatign -> incubating

wikier commented:
@silviulica, I've tried to address the version issue in the last commits. The new approach consists on reading the version from `pom.xml` during setup, and update `apache_beam.__version__` for later static usage with no other complex operations at runtime.

Thanks @aaltay, I think CI should be fine now with respect to this PR, but still fails because setup issue of the overall python tool suite. We may need to take a look to it with more time.

silviulica commented on sdks/python/setup.py:
```diff
@@ -17,29 +17,42 @@
-  return global_names['__version__']
+# Reads the actual version from pom.xml file, and synchronizes
+# apache_beam.__version__ field for later usage.
+def sync_version():
```
I do not think you need this if the get_version_from_distribution() really works. you can use get_version_from_pom in setup.py and use get_version_from_distribution() in all the other places. This way you avoid the somewhat fragile rewriting the __init_.py file.

wikier replied:
The problem is I didn't get `get_version_from_distribution()` really working out of `pip`-controlled environments. So, since it's not a pure `setuptools` feature, I'm more think about removing such function.

silviulica commented on sdks/python/setup.py:
```diff
@@ -60,17 +73,18 @@ def get_version():
     'pyyaml>=3.10',
-    ]
-
+    'lxml'
```
Please add it in sorted order.

silviulica commented on sdks/python/setup.py:
```diff
@@ -60,17 +73,18 @@ def get_version():
     url='https://beam.incubator.apache.org',
     download_url='TBD',
-    author='Apache Software Foundation',
+    author='Apache Beam (incubating)',
```
Thanks very much for all these corrections.

silviulica commented on sdks/python/setup.py:
```diff
@@ -79,7 +93,7 @@ def get_version():
     ]),
     install_requires=REQUIRED_PACKAGES,
-    package_data={'': ['*.pyx', '*.pxd']},
+    package_data={'apache_beam': ['*.pyx', '*.pxd', 'pom.xml']},
```
An easy way to test this kind of change (I never know if I get setup.py right :-) is to create a fresh virtual env and then run a python setup.py test -- does this land the Cython files and pom file int he proper places?

aaltay replied:
Also you can use tox to run the same tests as Travis using:

```
pip install tox
tox -e py27 -c sdks/python/tox.ini
```

silviulica commented on sdks/python/setup.py:
```diff
@@ -17,29 +17,42 @@
-  except ImportError:
+    # Windows doesn't always provide int64_t.
     cythonize = lambda *args, **kwargs: []
+else:
```
Sorry for the nit but the entire code base uses 2 spaces for indentation. I know that externally 4 spaces are used but we should keep it at 2 until we decide the change the entire code base to 4. So can you please keep it at 2?

silviulica commented on sdks/python/apache_beam/__init__.py:
```diff
@@ -65,17 +65,16 @@
 
-
-import apache_beam.internal.pickler
+__version__ = "0.2.0-incubating.dev"  # FIXME: PEP 440
```
Can you please add a comment about why you decided this instead of something else? 
So that people know how to bump it up? Also this line is basically rewritten by code in setup.py. It is worth a comment.

silviulica commented on sdks/python/apache_beam/version.py:
```diff
@@ -15,6 +15,20 @@
+
+
+def get_distribution_version():
+    return pkg_resources.get_distribution("apache_beam").version
```
Cool! Did not know this existed!

silviulica commented:
Hi Sergio! We are getting close here. Thanks very much for your effort. Besides the comments in line can you please explain what is the purpose of the pom.xml? Perhaps as a comment in the pom.xml itself. It is needed to integrate with Apache build/test processes but this is where my knowledge ends. It would be nice to have this comment for people to understand why it is needed especially since it is unusual for a Python package.

wikier commented:
@silviulica, the main purpuse of this PR is to help Beam to manage the entire source base with Maven. It's a detail I always like to care in multi-language projects, because it's very important for a proper overall management. 

In the end the approach I used it's not very intrusive: The python folks can just focus on the Python stuff; but the Maven guys will automatically run some tasks when they're building the whole project.

Please, let me know if you find useful the approach, to then continue polishing details.

Thanks for the feedback.

wikier commented:
@aaltay, I think I know why the CI is not happy. If I manually run the test from Maven (which is the end relies on `tox` as it was:

```
cd sdks/python
mvn test
...
cmdargs: [local('/home/wikier/tmp/beam/sdks/python/.tox/py27/bin/pip'), 'install', '-U', '--no-deps', '/home/wikier/tmp/beam/sdks/python/.tox/dist/apache-beam-sdk-0.2.0-incubating.dev.zip']

Processing ./.tox/dist/apache-beam-sdk-0.2.0-incubating.dev.zip
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 20, in <module>
      File "/tmp/pip-UqgCv_-build/setup.py", line 23, in <module>
        from apache_beam.version import get_version_from_pom
      File "apache_beam/__init__.py", line 75, in <module>
        from apache_beam import coders
      File "apache_beam/coders/__init__.py", line 18, in <module>
        from apache_beam.coders.coders import *
      File "apache_beam/coders/coders.py", line 37, in <module>
        import dill
    ImportError: No module named dill
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-UqgCv_-build
You are using pip version 7.1.2, however version 8.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

py27 installed: You are using pip version 7.1.2, however version 8.1.2 is available.,You should consider upgrading via the 'pip install --upgrade pip' command.,wheel==0.24.0
___________________________________ summary ____________________________________
ERROR:   py27: InvocationError: /home/wikier/tmp/beam/sdks/python/.tox/py27/bin/pip install -U --no-deps /home/wikier/tmp/beam/sdks/python/.tox/dist/apache-beam-sdk-0.2.0-incubating.dev.zip (see /home/wikier/tmp/beam/sdks/python/.tox/py27/log/py27-7.log)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.657 s
[INFO] Finished at: 2016-06-29T10:55:20+02:00
[INFO] Final Memory: 24M/981M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:exec (setuptools-test) on project beam-sdks-python: Command execution failed. Process exited with an error: 1 (Exit value: 1) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

I noticed the command uses `--no-deps` as argument. I have to admit I have not much experience with `tox`, but I guess that's why later we get a `ImportError: No module named dill`...

silviulica commented:
@wikier - Definitely the approach is good. I just wanted to leave a record of why we are doing it.  It is most likely obvious for people using Maven often but was not obvious for me :-) Thanks again for doing it.

aaltay commented:
@wikier I have reproduced the error you are seeing. 

The problem is you are importing and calling apache_beam.* packages before the completion of the setuptools.setup() call. That call is responsible for installing requirements (including dill). Anything prior to that call should be self contained should not depend on anything but whatever is available in a default Python installation.

I understand this creates a chicken and egg problem. I am not familiar with maven - python integrations. Do you know how other people approach this problem?

One option I can think of is to open and parse the xml file for the version string inside the setup.py without using lxml (e.g. a single string search for version). It may not be robust but it will address this problem.

dhalperi commented on sdks/python/apache_beam/__init__.py:
```diff
@@ -65,18 +65,19 @@
 if sys.version_info.major != 2:
   raise RuntimeError(
-      'Dataflow SDK for Python is supported only on Python 2.7. '
+      'Apache Bean SDK for Python is supported only on Python 2.7. '
```
Beam

wikier commented:
I have an idea how to solve the issue, @aaltay!

wikier commented:
@aaltay as I said, I have not much experience with `tox`... 

In a0a788ebaf7f96a5e2d1df50fc98339dfd6d238f I moved the `pom.xml` parsing to `setup.py`, adding `lxml` as dependency in `tox.ini`. But CI does not look happy with that :-|

(I could change to a dirtier parsing approach, but I'd like to clarify this issue before moving forward in any path)

aaltay commented:
@wikier

Tox runs sdist prior to installing dependencies. sdist is done by setup.py to create a source distribution. There is a way to skip this step by adding 'skipsdist = True' under '[tox]'. However I am strongly against doing this because, this would mean that our users will require lxml package to be installed prior to creating a source distribution and later for installing and testing. I would rather avoid any dependencies that could not be managed by setup.py itself.

I am fine with a dirtier parsing approach. Just to clarify, I assume this will involve some simple string searches rather a complex xml parsing.

On a separate note while I was trying to install lxml on one of my computers I got an error like this 'Could not find function xmlCheckVersion in library libxml2. Is libxml2 installed?' I think this it by itself a good reason to avoid this dependency. It is plausible that some of our users will end up seeing this error too.

wikier commented:
OK @aaltay, I went for a dirtier parsing (regex-based) and now CI is happy (at least Travis, Jenkins fails because RAT).

jbonofre commented:
R: @jbonofre

aaltay commented:
Thank you @wikier . This looks better

Jenkins is failing because tox is not installed: 

"[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:exec (setuptools-test) on project beam-sdks-python: Command execution failed. Cannot run program "tox" (in directory "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_MavenVerify/sdks/python"): error=2, No such file or directory -> [Help 1]"

Jenkins workers do not have tox package by default and it should be installed as part of the pom file. You may install to the user install directory using 'pip install tox --user'. If you know a better way of doing this with Maven please do so.

aaltay commented on sdks/python/setup.py:
```diff
@@ -17,29 +17,50 @@
   try:
-    # pylint: disable=g-statement-before-imports,g-import-not-at-top
-    from Cython.Build import cythonize
+      # pylint: disable=g-statement-before-imports,g-import-not-at-top
```
2 spaces please instead of 4. I noticed that our linter is skipping the setup.py file. I will fix that.

silviulica commented:
Hi Sergio,
I think I am basically in an LGTM state :-) Please one more thing besides the tests passing requirement -- The sync_version() function from setup.py. Rewriting code in setup.py is quote fragile. For instance it is not clear if it works at all for binary distributions for which apache_beam/__init__.py may not exist as a plain file. Better avoid this headache. I would be happier with having the XML parsing code (using regex) duplicated in a file (version.py or even __init__.py) and have a get_version() function. 

For the sake of making progress with this CL you can just leave a TODO in the setup.py that we want to move in this direction. Having a unified Maven based system is a very good achievement and I would rather not block this any longer.
Thanks!
Silviu

aaltay commented on sdks/python/pom.xml:
```diff
@@ -0,0 +1,108 @@
+
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
```
Python distribution will package this file and not any of its parents. Could you add a comment somewhere here that, in Python we use this to read the version string and and it is fine to package as a standalone file?

wikier replied:
Sure, I'll add a comment in the `MANIFEST.in` file.

aaltay commented on sdks/python/pom.xml:
```diff
@@ -0,0 +1,108 @@
+  <parent>
+    <groupId>org.apache.beam</groupId>
+    <artifactId>beam-sdks-parent</artifactId>
+    <version>0.2.0-incubating-SNAPSHOT</version>
```
A question here.

Would it be possible to define a version for this package independent of the parent version? In Python we have not released anything yet so this is not exactly our version. And in the future I assume Python and Java may have releases at different times. So it would be best not to share the same version between them.

If this is not possible, please just add a comment that this is required to be a shared version.

wikier replied:
Tes, you can do that in Maven.

But well, this PR assumes you want to have aligned the versions from all artifacts. If that's not what we want, we can drop the idea of syncing versions, which is what is causing most of the discussions.

francesperry replied:
In an early community [discussion](https://lists.apache.org/thread.html/3b201b523701df077bee1a916a8af8dbaf3b11c28aa83015f71dad93@1455032762@%3Cdev.beam.apache.org%3E), we agreed on aligning the versions:

> Finally, we propose the standard approach where the entire source code lives in each branch and is released concurrently. We’d like to avoid the case where individual modules are released on different cadences and are being maintained in different branches of the main repository. This is beneficial because we don’t need to worry as much about versioning some of the surfaces between components. However, for components that have a stable surface or across languages, we can relax this, as appropriate. Additionally, this can be relaxed for hotfixes and different SDK languages.

If we want to start a discussion about relaxing things, we should do so on the dev@ list for visibility.

wikier commented:
@silviulica thanks for the feedback. 

Summarizing, as I already replies to another comment from @aaltay, I need you to decide if in Beam you want:

a) Release all artifacs under a common version
b) Use different versions for each artifact

Because if you are more in the second scenario (e.g., the Python SDK will follow a different release cycle), it does make sense to keep the `pom.xml`-related code in the PR.

wikier commented:
BTW @aaltay, I think I'd prefer to have the `tox` installation in the CI configuration (as you have for Travis) rather than in the Maven build.

aaltay commented:
Thank you @wikier

I am fine with having tox in the CI configuration. Jenkins configuration configuration is not part of the source tree. @dhalperi may help you with setting that up. That said the disadvantage of not having it as a dependency in the maven build would be that our users (who wants to use maven) would need to install tox separately. (Having it installed in the Travis config is fine because, there is no user path for installing the sdk through Travis.)

This change is LGTM as longs as all tests become green (either by updating the Jenkins configuration or Maven build).

silviulica commented:
@wikier - with respect to your question about versions we want #2. Eventually (long-term) we may be able to release in sync but right now Java and Python are in different stages (Java - much more advanced).

dhalperi commented on sdks/python/pom.xml:
```diff
@@ -0,0 +1,123 @@
+            </configuration>
+          </execution>
+          <execution>
+            <id>ssetup-test</id>
```
what does `ssetup` stand for here?

wikier replied:
typo

dhalperi commented on sdks/python/pom.xml:
```diff
@@ -0,0 +1,123 @@
+              <executable>python</executable>
+              <arguments>
+                <argument>setup.py</argument>
+                <argument>clean</argument>
```
how comprehensive is this clean? e.g., does it delete virtualenvs, random files created by tox or lint, etc? I think we want to go as far as we can here.

wikier replied:
that's a good point, @dhalperi 

for the moment I'm just doing a `python setup.py clean`, but we could add any other clean tasks

dhalperi commented on sdks/python/pom.xml:
```diff
@@ -0,0 +1,123 @@
+                <argument>setup.py</argument>
+                <argument>build</argument>
+                <argument>--build-base</argument>
+                <argument>target/build</argument> <!--FIXME: ${maven.build.dir} is not expanded -->
```
I was under the impression that `maven.build.dir` is a Maven 1.x option that is no longer supported. Isn't it something like `project.build.directory` or so?

dhalperi commented on sdks/python/setup.py:
```diff
@@ -35,11 +35,32 @@
+    regex = '.*<parent>\s*<groupId>[a-z\.]+</groupId>\s*<artifactId>[a-z\-]+</artifactId>\s*<version>([0-9a-zA-Z\.\-]+)</version>.*'
+    pattern = re.compile(regex)
+    search = pattern.search(pom)
+    return search.group(1).replace("-SNAPSHOT", ".dev")  # TODO: PEP 440 and incubating suffix
```
hopefully if we're copying version from Maven pom then we get the incubating suffix for free?

wikier replied:
Yes, but `0.2.0-incubating.dev`  is not yet valid according to [PEP 440](https://www.python.org/dev/peps/pep-0440/). I'd need to check how other podlings releasing Python artifacts have applied this policy.

dhalperi commented on sdks/python/setup.py:
```diff
@@ -59,18 +80,18 @@ def get_version():
-    'pyyaml>=3.10',
-    ]
-
+    'pyyaml>=3.10'
```
i think we usually prefer to leave the `,` at the end of the list, so that lines can be added and removed easily without stressing over whether they're last in the file. Deliberate removal?

dhalperi commented on sdks/python/pom.xml:
```diff
@@ -0,0 +1,123 @@
+              <goal>exec</goal>
+            </goals>
+            <configuration>
+              <executable>tox</executable>
```
Above, `tox` was installed with the `--user` option. So `tox` is not likely to be on the user `PATH` by default. This is causing the Jenkins failures we still see; Travis is passing because presumably `tox` is installed globally on the machine.

Hmmmmmmmmmmmmmmmmmmmm. It's not obvious to me what to do here. Python folks?

dhalperi replied:
Maybe we can make `pip` install `tox`, its virtual envs, and all dependencies into the `target/` folder so that they will be properly cleaned up/

dhalperi commented:
Sergio, thank you so much for this PR. I had been overwhelmed by the idea of getting Python support in Maven, and have learned a lot from how you did it. Maven lifecycle is pretty cool!

Jenkins is still failing:

```
Caused by: java.io.IOException: Cannot run program "tox" (in directory "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_MavenVerify/sdks/python"): error=2, No such file or directory
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at java.lang.Runtime.exec(Runtime.java:620)
	at org.apache.commons.exec.launcher.Java13CommandLauncher.exec(Java13CommandLauncher.java:61)
```

I added a comment in the `pom.xml` why this is happening, but I have to do some research to understand the best solution.

wikier commented:
@silviulica, in ebe4d0eef420c6feb8e3b1815d65d72a9e4a0487 I reverted back to the old way to manage the version. I kept the utilities implemented just in case they could be useful in the future, but they're not used anymore in the current code.

Thanks @dhalperi, I've been using this approach in some other projects, because I think is good to have all modules integrated in the main build. In 0a193fa3693bcfe1dfa49b7cd6cec349d7fe41a2 I switched to `virtualenv` for providing `tox`. But that not make Jenkins happy, since `virtualenv` is not available either: 

```
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:exec (setup-test-virtualenv) on project beam-sdks-python: Command execution failed. Cannot run program "virtualenv" (in directory "/home/jenkins/jenkins-slave/workspace/beam_PreCommit_MavenVerify/sdks/python"): error=2, No such file or directory -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:exec (setup-test-virtualenv) on project beam-sdks-python: Command execution failed.
```

Unless anyone else has another idea, we have to get either `tox` or `virtualenv` installed in the ci boxes, otherwise the build would never work.

dhalperi commented:
Can we not install virtualenv just as we were installing tox?

The key might be as simple as updating where python installs its --user packages to be inside the build dir (`target/`). Then we can put the absolute paths to `tox`/`virtualenv` in the `pom.xml` so that they will work on any machine. https://pip.readthedocs.io/en/latest/user_guide/#user-installs

I think we should try to keep optimizing for normal users even more than we are for Jenkins.

aaltay commented:
Installing tox is enough. virtualenv is a required package for tox and they will be installed together.

Travis does not have tox or virutalenv installed on the machine. We install them in the --user mode also. They will not be on the user PATH, so it is important to call them with their absolute paths. That should be: '$HOME/.local/bin/tox' for Jenkins. However note that this is platform dependent and for example on osx it will be $HOME/Library/Python/2.7/bin/tox. 

The install path could be customized by setting the PYTHONUSERBASE environment variable. I have tried this only on linux and I am not sure how it will behave on other systems.

export PYTHONUSERBASE=/somedirectory
pip install --user tox
/somedirectory/bin/tox ...

Cleanup of this is also easier. If we can manage to install things in a directory of our own choice we can just rm it at the cleanup phase.

wikier commented:
@aaltay do `pip install tox` with `--user` sounds good to be (also needs `--ignore-installed` to work on those environments where it's already installed). 

But if you take a look to the lasts commits I'm fighting against the Maven version at Jenkins, since it keeps trying to execute the goal `exec:java` when we only want `exec:exec`, and works locally and at Travis. I've lost too much time on that, and I'm running out of ideas what could actually work there :disappointed:

dhalperi commented on pom.xml:
```diff
@@ -226,6 +226,9 @@
                   <exclude>.repository/**/*</exclude>
+                  <exclude>**/.tox/**/*</exclude>
+                  <exclude>**/*.egg-info/**</exclude>
+                  <exclude>**/.eggs/**</exclude>
```
please put all of these in a directory that is automatically cleaned by Maven, like `/target/`

aaltay commented:
@wikier. I have spent some time trying to figure out what is happening to no luck. I have been hitting into the same problem you mentioned. I am not a maven expert, I am guessing it is tripping off when it is trying to execute verify phase from the parent pom.xml where a partially define exec:java goal is used.. Presumably this change will require some changes there too.

At this point you can either do a partial submit. By removing the parent dependency in pom.xml and therefore not triggering jenkins runs at all. Or shelve this change for now. 

I can spend more time on this in about 2 weeks. Depending on your preference I can take over this change or help you fighting the maven. I would also appreciate if anybody with more maven knowledge could give us a hand.

wikier commented:
@aaltay, @dhalperi, the problem is that I can't reproduce why Jenkins executes the Maven goal `exec:java`...

dhalperi commented:
@wikier : I believe the java goal is coming [from the root pom](https://github.com/apache/incubator-beam/blob/master/pom.xml#L760):

```xml
        <plugin>
          <groupId>org.codehaus.mojo</groupId>
          <artifactId>exec-maven-plugin</artifactId>
          <version>1.4.0</version>
          <executions>
            <execution>
              <phase>verify</phase>
              <goals>
                <goal>java</goal>
              </goals>
            </execution>
          </executions>
          <configuration>
            <cleanupDaemonThreads>false</cleanupDaemonThreads>
            <systemProperties>
              <systemProperty>
                <key>java.util.logging.config.file</key>
                <value>logging.properties</value>
              </systemProperty>
            </systemProperties>
          </configuration>
        </plugin>
```

wikier commented:
@dhalperi, yes, I know the `exec:java` comes from the parent `pom.xml`, but I still don't understand why is being call in Jenkins and not in Travis or my local environment...

sergioisidoro commented:
Hi, Sorry to intrusion. 
I was curious and decided to build this out of curiosity

```
/Users/sergioisidoro/Development/sandboxes/beam/lib/python2.7/site-packages/google_apitools-0.5.4-py2.7.egg/apitools/base/protorpclite/messages.py in __init__(cls, name, bases, dct)
    301     def __init__(cls, name, bases, dct):
    302         # Can only define one level of sub-classes below Enum.
--> 303         if not (bases == (object,) or bases == (Enum,)):
    304             raise EnumDefinitionError(
    305                 'Enum type %s may only inherit from Enum' % name)

NameError: global name 'Enum' is not defined
```

Further check and it appears that the problem might be this:

https://github.com/google/apitools/issues/123
https://github.com/google/apitools/blob/master/apitools/base/protorpclite/messages.py#L303

aaltay commented:
Hi @sergioisidoro I am confused about your comment, could you please explain it a bit more? The problem we are experiencing now is Maven related. We are having trouble getting the `exec` rule running without inheriting from the parents. And we have not seen the error you mentioned: `global name 'Enum' is not defined`.

aaltay commented:
@wikier I can reproduce this locally by:

```
cd sdks/python
mvn clean install
```

I got the same error as Jenkins. I have Maven 3.0.5 installed. Do you not get any error when you run the above locally? Maybe it is related to the Maven version.

It does not happen in Travis, because it does not use Maven to build.

wikier commented:
With Maven ` 3.3.9` I guess a different behavior, `the goal ` exec:java` is not executed.

aaltay commented:
@wikier I have tried this with mvn `3.3.9` and mojo plugin `1.4.0` and `1.5.0` and I can still reproduce.

I managed to get this running by disabling the inheritance on the root pom.xml. Adding `<inherited>false</inherited>` in the root pom.xml as below fixes the problem.
```
...
          <version>1.4.0</version>
          <inherited>false</inherited>
          <executions>
...

```

@dhalperi , is this an acceptable solution? Would it introduce problems with other pom files that needs to inherit from the root?

wikier commented:
@aaltay in Maven by default plugin configuration is propagated to child modules ([`inherited=true`](https://maven.apache.org/guides/mini/guide-configuring-plugins.html#Using_the_inherited_Tag_In_Build_Plugins)), so I guess that would break other child modules who depend on the `exec:java` configuration form the parent pom.

The real solution would be to [skip `exec:java`](http://www.mojohaus.org/exec-maven-plugin/java-mojo.html#skip) just on the Python SDK, but I didn't manage to get it working in my previous tries. I'll give it another shot...

wikier commented:
Now I got that working :smile:

But the build still fails because `pylint` issues:

```
OK
py27 runtests: commands[2] | /home/jenkins/jenkins-slave/workspace/beam_PreCommit_MavenVerify/sdks/python/run_pylint.sh
Running pylint on changed files:
apache_beam/__init__.py
apache_beam/internal/apiclient.py
apache_beam/utils/dependency.py
apache_beam/utils/dependency_test.py
apache_beam/version.py
setup.py
************* Module apache_beam.internal.apiclient
W: 33, 0: Unused pickler imported from apache_beam.internal (unused-import)
W: 40, 0: Unused get_sdk_name_and_version imported from apache_beam.utils.dependency (unused-import)
ERROR: InvocationError: '/home/jenkins/jenkins-slave/workspace/beam_PreCommit_MavenVerify/sdks/python/run_pylint.sh'
___________________________________ summary ____________________________________
ERROR:   py27: commands failed
```

I'll take care..

wikier commented:
@aaltay, @dhalperi I got solve the issue we have been discussed, finally!

Now RAT is not happy; how can I retrieve the report generated by the Jenkins build?

aaltay commented:
@wikier Great news. I will start reviewing the new changes.

I ran the same mvn locally similar to Jenkins to generate the RAT report using:

`mvn -B -e -P release help:effective-settings clean install coveralls:report -DskipITs=true`

RAT is complaining about files under `nose-1.3.7-py2.7.egg/*`

I pasted the output to this gist: [beam-sdks-python-0.2.0-incubating-SNAPSHOT.rat](https://gist.github.com/aaltay/ebb36cb5eaaf554e6b43ca8f71b9cd40)

aaltay commented on sdks/python/apache_beam/__init__.py:
```diff
@@ -63,20 +63,16 @@
-import sys
-
-
-if sys.version_info.major != 2:
```
I believe it is important to make this check before the apache_beam import. This would avoid an import related error due to and old Python version happening before we got a chance to raise this version error.

aaltay commented on sdks/python/apache_beam/utils/dependency.py:
```diff
@@ -58,15 +58,14 @@
 import tempfile
 
-
+from apache_beam.version import __version__
```
It looks like this import was superfluous.  This is the one being used:

`from apache_beam import version as beam_version`

Can we remove `from apache_beam.version import __version__`?

aaltay commented on sdks/python/apache_beam/internal/apiclient.py:
```diff
@@ -117,7 +117,7 @@ def __init__(self, packages, options, environment_version):
     self.local = 'localhost' in self.google_cloud_options.dataflow_endpoint
 
-    sdk_name, version_string = get_sdk_name_and_version()
+    version_string = version.__version__
```
This reversion looks like a bad merge.

`get_sdk_name_and_version()` would return `version.__version__` and it will also use it to set the sdk name correctly.

aaltay commented:
I added some comments, but I could not do a full review. There are some bad merges and a long history by now. For example there are two `def get_version():` in `setup.py`.

Could you do a rebase and merge with python-sdk branch?

wikier commented:
@aaltay I have addressed most of the comments you have made. Thanks for your review.

Unfortunately I can't run `mvn -B -e -P release help:effective-settings clean verify coveralls:report -DskipITs=true` because of course I have no rights to deploy:

```
py27 runtests: commands[2] | /home/wikier/tmp/beam/sdks/python/run_pylint.sh
Enter passphrase for key '/home/wikier/.ssh/id_rsa': 
Received disconnect from 192.30.253.113 port 22:11: Bye Bye
Connection to github.com closed by remote host.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
ERROR: InvocationError: '/home/wikier/tmp/beam/sdks/python/run_pylint.sh'
```

But I least I got far enough to get the RAT details. So now Jenkins looks happy with the `beam-sdks-python` :smile:. But I think I've introduce a lateral effect fighting against the `exec:java` issue, because later `beam-runners-flink_2.10-examples` fails:

```
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:java (beam-exec-java) on project beam-runners-flink_2.10-examples: Unable to parse configuration of mojo org.codehaus.mojo:exec-maven-plugin:1.4.0:java for parameter arguments: Cannot store value into array: ArrayStoreException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:java (beam-exec-java) on project beam-runners-flink_2.10-examples: Unable to parse configuration of mojo org.codehaus.mojo:exec-maven-plugin:1.4.0:java for parameter arguments: Cannot store value into array
```

I'll take a look how I can reproduce that locally.

wikier commented:
I believe the following error is not related with this PR:

```
Exception in thread "main" java.lang.IllegalArgumentException: No Runner was specified and the DirectRunner was not found on the classpath.
Specify a runner by either:
    Explicitly specifying a runner by providing the 'runner' property
    Adding the DirectRunner to the classpath
    Calling 'PipelineOptions.setRunner(PipelineRunner)' directly
	at org.apache.beam.sdk.options.PipelineOptions$DirectRunner.create(PipelineOptions.java:282)
	at org.apache.beam.sdk.options.PipelineOptions$DirectRunner.create(PipelineOptions.java:273)
	at org.apache.beam.sdk.options.ProxyInvocationHandler.getDefault(ProxyInvocationHandler.java:481)
	at org.apache.beam.sdk.options.ProxyInvocationHandler.invoke(ProxyInvocationHandler.java:156)
	at org.apache.beam.sdk.options.PipelineOptionsValidator.validate(PipelineOptionsValidator.java:71)
	at org.apache.beam.sdk.options.PipelineOptionsFactory$Builder.as(PipelineOptionsFactory.java:306)
	at org.apache.beam.runners.flink.examples.WordCount.main(WordCount.java:107)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
```

Right?

aaltay commented:
@wikier I do not have rights to deploy either so I am not sure why you got the error. But that is ok as long as the RAT issue is resolved.

The error you got seems like related to the #724. But this is very confusing to me. Because we have not merged this code the `python-sdk` branch yet. If you look at the branch the failing example does not exist. It seems like you pulled from `master` and `python-sdk` branches at different times. Could you clean up your change a little bit? Your changes touch to a few files, we need to apply those changes on top of a clean `python-sdk` branch.

wikier commented:
@aaltay I always merged from `upstream/python-sdk`, actually the branch is up-to-date. Therefore the bug should be there where you created the branch, and now we see it here because the inclusion of the Python SDK. Do you plan to merge `master` into `python-sdk`, so we can get issue #724 also fixed here?

aaltay commented:
@wikier My mistake, I agree branch seems in-sync with python-sdk.

I am still confused, because Jenkins pre-commit is passing for other jobs with the existing `runners/flink/examples/pom.xml` (for ex: #765). I am not sure why it is failing here.

Nevertheless we should merge `master`. We don't have a plan yet but I will update here once there is one. In the meantime could you take another look at that file, are we sure that this will be fixed when we do the merge?

aaltay commented on sdks/python/setup.py:
```diff
@@ -85,7 +85,6 @@ def get_version():
         'apache_beam/utils/counters.py',
         'apache_beam/utils/windowed_value.py',
     ]),
-    setup_requires=['nose>=1.0'],
```
Why is this removed? This is needed for a user to run `python setup.py test`.

wikier replied:
@aaltay the problem is that `setup_requires` gets the dependency in `./.eggs/`, which was causing problems with RAT about unlicensed files. That's why I move to manage it with `tox`.

If you know any way to customize that path (e.g., `./target/eggs/`) then I'm happy to revert it.

aaltay replied:
We need to find an alternative solution. Because, people getting just the python-sdk part and not wanting to use maven would need this so that `nose` would be installed when they run `python setup.py test`

Would be possible to tell RAT to ignore certain folders.

aaltay replied:
Excluding the nose folder in the pom.xml from RAT fixed this for me: https://github.com/aaltay/incubator-beam/commit/752d689335e606e3ef1ce40e07e8f89b59eafed3

wikier commented:
@aaltay about the Flink runner example, I can;t warranty. But I'd prefer, if possible, that you merge `master` into `python-sdk` to actually check it before merging this PR.

aaltay commented:
@wikier  #787 merged master into python-sdk. Please update and give it another try. Hopefully Jenkins issue will be fixed now.

wikier commented:
Merged from upstream, @aaltay. But Jenkins build behaved weird locally resolving parent poms:

```
java.io.IOException: remote file operation failed: /home/jenkins/jenkins-slave/workspace/beam_PreCommit_MavenVerify at hudson.remoting.Channel@e34346d:beam2: hudson.remoting.ProxyException: hudson.maven.MavenModuleSetBuild$MavenExecutionException: org.apache.maven.project.ProjectBuildingException: Some problems were encountered while processing the POMs:
[FATAL] Non-resolvable parent POM: Could not find artifact org.apache.beam:beam-parent:pom:0.3.0-incubating-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 22, column 11
[FATAL] Non-resolvable parent POM for org.apache.beam:beam-runners-parent:0.3.0-incubating-SNAPSHOT: Could not find artifact org.apache.beam:beam-parent:pom:0.3.0-incubating-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ org.apache.beam:beam-runners-parent:0.3.0-incubating-SNAPSHOT, /home/jenkins/jenkins-slave/workspace/beam_PreCommit_MavenVerify/runners/pom.xml, line 22, column 11
[FATAL] Non-resolvable parent POM: Could not find artifact org.apache.beam:beam-parent:pom:0.3.0-incubating-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 22, column 11
[FATAL] Non-resolvable parent POM for org.apache.beam:beam-sdks-parent:0.3.0-incubating-SNAPSHOT: Could not find artifact org.apache.beam:beam-parent:pom:0.3.0-incubating-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ org.apache.beam:beam-sdks-parent:0.3.0-incubating-SNAPSHOT, /home/jenkins/jenkins-slave/workspace/beam_PreCommit_MavenVerify/sdks/pom.xml, line 22, column 11
[FATAL] Non-resolvable parent POM: Could not find artifact org.apache.beam:beam-parent:pom:0.3.0-incubating-SNAPSHOT and 'parent.relativePath' points at wrong local POM @ line 22, column 11
```

Locally for me `mvn -B -e -P release help:effective-settings clean verify coveralls:report -DskipITs=true` get over that and normally continues the build.

wikier commented:
@aaltay, after merging `python-sdk` from upstream, we still have the same issue in this branch:

```[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:java (beam-exec-java) on project beam-runners-flink_2.10-examples: The parameters 'mainClass' for goal org.codehaus.mojo:exec-maven-plugin:1.4.0:java are missing or invalid -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.4.0:java (beam-exec-java) on project beam-runners-flink_2.10-examples: The parameters 'mainClass' for goal org.codehaus.mojo:exec-maven-plugin:1.4.0:java are missing or invalid```

I've tried to get rid of the `java` goal in `beam-runners-flink_2.10-examples` with the same trick I'm using in `beam-sdks-python`, but it does not work... Maybe we could involve the guys from #724 (@mxm) to figure out what's the lateral issue here.

aaltay commented:
@wikier

I got both python-sdk and flink examples running locally. See this commit: https://github.com/aaltay/incubator-beam/commit/a1df1f57f14261edf34c5ae2efa3f88d7b8a8225

It is based on yours. Except that I used <pluginManagement> to disable execution. I noticed the flink was using this too. And adding an id there broke their disable mechanism. With this both places can use the same way of disabling.

(In this commit I changed 1 place related to maven version to run locally with my old maven installation. That part should not be submitted.)

Could you try this please?

aaltay commented:
I commented on the RAT issue as well. Hopefully applying both will pass all the tests. Fingers crossed. @wikier Thank you again.

wikier commented:
Cool, @aaltay! That trick works better, actually avoiding inheritance in `exec-maven-plugin` executions previously configure.

Also reverted `nose` to `setup_requires`, and properly configured RAT exclusion.

So now the CI system should be, finally, happy :smile: 

Thanks so much for your time with these last details. I hope the overall idea of this PR would be useful for the podling.

wikier commented:
@aaltay I've patch the RAT configuration, but Jenkins reports unknown licenses I'm not able to reproduce locally...

aaltay commented:
@wikier

The error is because `nose-1.3.7-py2.7.egg` folder is created directly under `python` and not under `.egg` that is why RAT is still complaining about nose related files.

I cloned your branch and make this single line change: https://github.com/apache/incubator-beam/pull/814/commits/232590ed7cdaddc133796ee86f1084d12e4b368a

It fixes the problem. I also made a temporary PR (#814) to test this with the actual thing Jenkins is running, and the tests passed.

I believe if you apply this change, all tests will pass.

wikier commented:
Looks like the version of `setuptools` I have locally installed and the one available in Jenkins differ in that behavior. Thanks for the pointed, @aaltay!

So, finally, this PR is ready for discussion :smile:. Should I bring it to [dev@beam](https://lists.apache.org/list.html?dev@beam.apache.org) or better someone from the core team?

aaltay commented:
LGTM.

R: @dhalperi for a final review.

Thank you @wikier. It is great that your change gives us integration with maven.

That said there are a few issues with the version management part. These should NOT block the merging of this PR. I will work on a follow up PR right after this is merged to address those.

As a note the issues are `get_version_from_pom` and `sync_version` are not called from anywhere. And I will fix the code to call them in my PR.

wikier commented:
Perfect. Let me know if you need anything else, @dhalperi.

Back to the version issue @aaltay, if you read the thread above, I initially proposed to align versions. Then the feedback I got from @silviulica was that you didn't want to do that for now. So I kept the helper functions I implemented for whenever such issues is going to be revisited. I'd handle that in a separate PR, I've created [BEAM-547](https://issues.apache.org/jira/browse/BEAM-547) for that.

aaltay commented:
@wikier Sounds good. I remember it was discussed earlier in the thread. Let's get this PR merged and let's move the version discussion to the https://issues.apache.org/jira/browse/BEAM-547. Thank you for opening that issue.

dhalperi commented:
Merged manually, please close PR at your leisure @wikier 

@aaltay is going to fixup some things I broke during the merge. But it should be okay.

Thanks again Sergio!

wikier commented:
Ah ok, merged in 76de6109fa19db34390edc593a62524a92f54ca0. You are welcome ;-)
